File: PDI/dano.js
const express = require("express");
const mysql = require('mysql');

async function dano(IpPath, IrPath) {
    const { createCanvas, loadImage } = require('canvas');
    const fs = require('fs');

    // Leer las imágenes
    const Ip = await loadImage(IpPath);
    const Ir = await loadImage(IrPath);

    // Umbral para el canal rojo y verde
    let rThr, gThr, bThr;

    // Función para extraer el canal rojo y verde y contar píxeles
    const countColorPixels = (image, rThreshold, gThreshold, bThreshold) => {
        const canvas = createCanvas(image.width, image.height);
        const ctx = canvas.getContext('2d');
        ctx.drawImage(image, 0, 0);
        const imageData = ctx.getImageData(0, 0, canvas.width, canvas.height);
        const data = imageData.data;
        let colorPixelCount = 0;

        for (let i = 0; i < data.length; i += 4) {
            const r = data[i];
            const g = data[i + 1];
            const b = data[i + 2];
            const mask = new Array(image.width * image.height).fill(false);

            if (r >= rThreshold[0] && r <= rThreshold[1] &&
                g >= gThreshold[0] && g <= gThreshold[1] &&
                b >= bThreshold[0] && b <= bThreshold[1]) {
                mask[i / 4] = true;
            }
        }

        return colorPixelCount;
    };

    // Contar píxeles rojos en las imágenes
    const IpRedPix = countColorPixels(Ip, [0, 255], [0, 140], [0, 140]);
    const IrRedPix = countColorPixels(Ir, [0, 255], [0, 140], [0, 140]);

    // Contar píxeles verdes en las imágenes
    const IpGreenPix = countColorPixels(Ip, [0, 255], [0, 0], [0, 110]);
    const IrGreenPix = countColorPixels(Ir, [0, 255], [0, 0], [0, 110]);

    // Crear imágenes solo con el canal rojo y verde
    const createColorImage = (image, redMask, greenMask) => {
        const canvas = createCanvas(image.width, image.height);
        const ctx = canvas.getContext('2d');
        ctx.drawImage(image, 0, 0);
        const imageData = ctx.getImageData(0, 0, canvas.width, canvas.height);
        const data = imageData.data;

        for (let i = 0; i < data.length; i += 4) {
            data[i] = redMask[i / 4] ? data[i] : 0;
            data[i + 1] = greenMask[i / 4] ? data[i + 1] : 0;
            data[i + 2] = 0;
        }

        ctx.putImageData(imageData, 0, 0);
        return canvas.toBuffer();
    };

    // Crear imágenes solo con el canal rojo y verde
    const IpDano = createColorImage(Ip, IpRedPix, IpGreenPix);
    const IrDano = createColorImage(Ir, IrRedPix, IrGreenPix);

    // Calcular porcentaje de celdas dañadas
    const porcentajeDanadas = Math.round(((IrGreenPix / IpGreenPix) - 1) * 100);

    return { IpDano, IrDano, porcentajeDanadas };
}

File: PDI/dano.m
function [IpDano, IrDano, porcentajeDanadas] = dano(IpPath, IrPath)
    Ip = imread(IpPath);
    Ir = imread(IrPath);

    % Extraer canal rojo de imagen patrón
    rThr = 0;
    gThr = 140;
    bThr = 140;

    IpRed = Ip(:,:,1) >= rThr & Ip(:,:,2) <= gThr & Ip(:,:,3) <= bThr;

    % Extraer canal rojo de imagen para analizar
    IrRed = Ir(:,:,1) >= rThr & Ir(:,:,2) <= gThr & Ir(:,:,3) <= bThr;
    
    % Extraer canal verde de imagen patrón
    rThr = 255;
    gThr = 0;
    bThr = 110;

    IpGreen = Ip(:,:,1) <= rThr & Ip(:,:,2) >= gThr & Ip(:,:,3) <= bThr;
    IpGreenPix = sum(sum(IpGreen));
    
    % Extraer canal verde de imagen para analizar
    IrGreen = Ir(:,:,1) <= rThr & Ir(:,:,2) >= gThr & Ir(:,:,3) <= bThr;
    IrGreenPix = sum(sum(IrGreen));
    
    % Imágenes solo con el canal verde
    blk = zeros(size(Ip,1), size(Ip,2), 'uint8');
    IpDano = uint8(cat(3,IpRed,IpGreen,blk)).*Ip;
    IrDano = uint8(cat(3,IrRed,IrGreen,blk)).*Ir;
    
    % Cálculo de porcentaje de celdas tapadas
    porcentajeDanadas = uint8((((IrGreenPix) / (IpGreenPix)) - 1) * 100);
end